google/cloud/pubsublite/internal/partition_publisher.cc

```diff
@@ -0,0 +1,353 @@
+      cancel_token_{alarm->RegisterAlarm(
+          absl::Milliseconds(alarm_period_ms),
+          std::bind(&PartitionPublisherImpl::OnAlarm, this))} {}
+
+PartitionPublisherImpl::~PartitionPublisherImpl() {
+  future<void> shutdown = Shutdown();
+  if (!shutdown.is_ready()) {
+    GCP_LOG(WARNING) << "`Shutdown` must be called and finished before object "
+                        "goes out of scope.";
+    assert(false);
+  }
+  shutdown.get();
+}
+
+future<Status> PartitionPublisherImpl::Start() {
+  promise<void> root;
+  future<void> root_future = root.get_future();
+  future<Status> return_future;
+  {
+    std::lock_guard<std::mutex> g{mu_};
+    root_future.then(ChainFuture(resumable_stream_->Start()))
+        .then([this](future<Status> start_status) {
+          bool has_shutdown = true;
+          promise<Status> start_promise{null_promise_t{}};
+          {
+            std::lock_guard<std::mutex> g{mu_};
+            if (!start_future_) return;
+            if (!shutdown_) {
+              has_shutdown = false;
+              shutdown_ = true;
+            }
+            start_promise = std::move(*start_future_);
+            start_future_.reset();
+          }
+          Status status = start_status.get();
+          start_promise.set_value(status);
+          if (!has_shutdown) {
+            cancel_token_ = nullptr;
+            SatisfyOutstandingMessages(status);
+          }
+        });
+  }
+  root.set_value();
+  Read();
+  return start_future_->get_future();
+}
+
+future<StatusOr<Cursor>> PartitionPublisherImpl::Publish(PubSubMessage m) {
+  future<StatusOr<Cursor>> message_future;
+  {
+    std::lock_guard<std::mutex> g{mu_};
+    if (shutdown_) return make_ready_future(StatusOr<Cursor>(Status()));
+    promise<StatusOr<Cursor>> message_promise;
+    message_future = message_promise.get_future();
+    unbatched_messages_with_futures_.emplace_back(
+        MessageWithFuture{std::move(m), std::move(message_promise)});
+  }
+  return message_future;
+}
+
+void PartitionPublisherImpl::Flush() {
+  {
+    std::lock_guard<std::mutex> g{mu_};
+    if (writing_ || shutdown_) return;
+    writing_ = true;
+  }
+  Rebatch(false);
+  WriteBatches();
+}
+
+future<void> PartitionPublisherImpl::Shutdown() {
+  promise<void> root;
+  future<void> root_future = root.get_future();
+  {
+    std::lock_guard<std::mutex> g{mu_};
+    if (shutdown_) return make_ready_future();
+    shutdown_ = true;
+    root_future = root_future.then(ChainFuture(resumable_stream_->Shutdown()))
+                      .then([this](future<void>) {
+                        SatisfyOutstandingMessages(Status());
+                      });
+  }
+  cancel_token_ = nullptr;
+  root.set_value();
+  return root_future;
+}
+
+void PartitionPublisherImpl::Rebatch(bool process_in_flight) {
+  std::lock_guard<std::mutex> g{mu_};
+  unsent_batches_ =
+      CreateBatches(UnbatchAllLockHeld(process_in_flight), batching_settings_);
+}
```

Comment:
Please 1) inline this and 2) remove the boolean parameter from UnbatchAllLockHeld. The ones which call this with `false` should append to unsent_batches_ not overwrite it.